tarxx.h

```diff
@@ -101,20 +104,102 @@ namespace tarxx {
 
         void add_file(const std::string& filename)
         {
+            if (stream_file_header_pos_ >= 0) throw std::logic_error("Can't add new file before add_file_complete has been called");
+
             block_t block;
             std::fstream infile(filename, std::ios::in|std::ios::binary);
             if (!infile.is_open()) return;
             write_header(filename);
             while (infile.good()) {
                 infile.read(block.data(), block.size());
-                auto read = infile.gcount();
+                const auto read = infile.gcount();
                 if (read < block.size()) std::fill_n(block.begin() + read, block.size() - read, 0);
                 write(block);
             }
         }
 
+        void add_file_streaming()
+        {
+            if (stream_file_header_pos_ >= 0) throw std::logic_error("Can't add new file before add_file_complete has been called");
+            if (mode_ != output_mode::file_output) throw std::logic_error(__func__ + " only supports output mode file"s);
+
+            // write emtpy header
+            stream_file_header_pos_ = file_.tellg();
+            block_t header {};
+            write(header);
+        }
+
+        void add_file_streaming_data(const char* const data, std::streamsize size)
+        {
+            unsigned long pos = 0;
+
+            // fill a new block with old and new data if enough data
+            // is available to fill a whole block
+            if ((stream_block_used_ + size) >= BLOCK_SIZE) {
+                block_t block;
+
+                // copy saved data
+                std::copy_n(stream_block_.data(), stream_block_used_, block.data());
+
+                // copy from new data
+                const auto copy_from_new_data = BLOCK_SIZE - stream_block_used_;
+                std::copy_n(data, copy_from_new_data, block.data() + stream_block_used_);
+
+                // write new complete block
+                write(block);
+
+                pos += copy_from_new_data;
+                size -= copy_from_new_data;
+                stream_block_used_ = 0;
+            }
+
+            // write new block as long as we have enough data to fill a whole block
+            while (size >= BLOCK_SIZE) {
+                block_t block;
```

Comment:
I've not thought this through yet, but it maybe, possibly would be nice to allocate one `block_t` only and reuse it (share the allocation betwen lines 139 and 158). also, if the compiler isn't smart enough it would maybe re-allocate (and re-initialize?) the block for every iteration. I'd suggest to rather extract the allocation and share it throughout the algorithm.

Reply:
I moved block to the top so it's reused. shouldn't do any harm as the block wasn't initialized anyways and we're only using defined ranges so I'm not 0ing anything now either